### PR TITLE
Simple React app

### PR DESCRIPTION
**Pull request by rrrds**
### React Basics

[Demo](https://rrrds.github.io/simple-react-app/) |
[Code base](https://github.com/rrrds/simple-react-app)

The code is submitted in a dedicated feature branch.
Please, review.

### CODE REVIEW

**Comment by zonzujiro** [comment]
You have missing functionality - I don't see search, sorting and filtration.

File: submissions/rrrds/simple-react/App/App.js [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+import { Character } from "../Character";
+import LocationList from "../LocationList";
+import { Location } from "../Location";
+import EpisodeList from "../EpisodeList";
+import { Episode } from "../Episode";
+
+function App() {
+  return (
+    <Router>
+      <div className="app">
```
Comment (zonzujiro) [inline submissions/rrrds/simple-react/App/App.js]: Looks like redundant `div`

File: submissions/rrrds/simple-react/Character/index.js [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+  }, []);
+
+  useEffect(() => {
+    if (!character) {
+      return;
+    }
+
+    const episodesId = character.episode.map(ep => ep.id);
+    fetchEpisode(episodesId).then(data => {
+      if (Array.isArray(data)) {
```
Comment (zonzujiro) [inline submissions/rrrds/simple-react/Character/index.js]: Why it can be not an array?

Comment (rrrds) [reply submissions/rrrds/simple-react/Character/index.js]: If character participate only in one episode, API will return single object fit episode info

File: submissions/rrrds/simple-react/Character/index.js [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+    }
+
+    const episodesId = character.episode.map(ep => ep.id);
+    fetchEpisodeMulti(episodesId).then(data => {
+      setEpisodes(data);
+    });
+  }, [character]);
+
+  return (
+    character && (
```
Comment (zonzujiro) [inline submissions/rrrds/simple-react/Character/index.js]: It is a lot of lines to render. I believe better will be if you handle this with `return`

File: submissions/rrrds/simple-react/api.js [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+    .join("&");
+
+  return await fetchData(`/character/?page=${page}&${filterString}`);
+}
+
+export async function fetchEpisodeAll(page = 1, filters) {
+  const allFilters = { name: "" };
+  filters = Object.assign(allFilters, filters);
+
+  const filterString = Object.entries(filters)
```
Comment (zonzujiro) [inline submissions/rrrds/simple-react/api.js]: Pls, use `URLSearchParam` instead

File: submissions/rrrds/simple-react/api.js [hunk 5]
```diff
@@ -0,0 +1,135 @@
+const API_BASE = "https://rickandmortyapi.com/api";
+
+function addIds(data, propNames) {
+  propNames = [].concat(propNames || []);
```
Comment (zonzujiro) [inline submissions/rrrds/simple-react/api.js]: Pls, use the default function arguments instead `|| []`

File: submissions/rrrds/simple-react/api.js [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+
+  const filterString = Object.entries(filters)
+    .map(([key, value]) => {
+      return `${key}=${value}`;
+    })
+    .join("&");
+
+  return await fetchData(`/location?page=${page}&${filterString}`).then(
+    data => {
+      console.log(data);
```
Comment (zonzujiro) [inline submissions/rrrds/simple-react/api.js]: `console.log` is unacceptable in prod code

File: submissions/rrrds/simple-react/api.js [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+    return [].concat(data || []).map(item => addIds(item, "residents"));
+  });
+
+  return res;
+}
+
+async function fetchData(path) {
+  return await fetch(`${API_BASE}${path}`)
+    .then(res => {
+      if (res.status !== 200) return null;
```
Comment (zonzujiro) [inline submissions/rrrds/simple-react/api.js]: `201`, `204`?

File: submissions/rrrds/simple-react/api.js [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+export async function fetchLocation(id) {
+  const res = await fetchData(`/location/${id}`).then(data =>
+    addIds(data, "residents")
+  );
+
+  return res;
+}
+
+export async function fetchLocationMulti(id) {
+  const idStr = [].concat(id || []).join(",");
```
Comment (zonzujiro) [inline submissions/rrrds/simple-react/api.js]: You can send a request with no ids in such case. Better will be not to send a request at all.

**Pull request merged**